### PR TITLE
Update RedundantBegin cop to respect heredocs

### PR DESCRIPTION
**Pull request by thirtysixthspan**
The RedundantBegin cop does not respect heredocs, stripping all but the first line of the heredoc when updating the source. This PR adds failing tests to demonstrate. 

This PR also updates the strategy for removing the extra begin/end. In the previous version, when a heredoc was used in the last command of the block, the end of the range to be replaced was returned as the first line of the heredoc, not the last line of the heredoc. I believe this may be an issue in the parsing library not correctly returning ranges for heredoc. The new strategy is a work around until the root issue can be resolved.

### CODE REVIEW

File: lib/rubocop/cop/style/redundant_begin.rb [hunk 1]
```diff
@@ -46,9 +46,11 @@ def autocorrect(node)
 
             corrector.replace(
               range_with_surrounding_space(node.loc.expression),
-              range_with_surrounding_space(
-                child.loc.expression
-              ).source.gsub(/^[ \t]{#{indent_diff}}/, '')
+              range_with_surrounding_space(node.loc.expression)
+                .source
+                .gsub(/\A\s*?begin\s*?(;|$)/, '')
```
Comment (bbatsov) [inline lib/rubocop/cop/style/redundant_begin.rb]: Too many regexps are a recipe for bugs IMO. It's always preferable to manipulate the nodes directly. @jonas054 Thoughts?

Comment (thirtysixthspan) [inline lib/rubocop/cop/style/redundant_begin.rb]: Agreed. My first two failed attempts to eliminate this destructive behavior manipulated the nodes directly. Maybe someone with more knowledge of the parsing library, where I believe the root of the problem lies, could suggest an alternative solution?

Comment (jonas054) [inline lib/rubocop/cop/style/redundant_begin.rb]: It's difficult to come up with a simple, elegant solution. I have a possibly acceptable suggestion [here](https://gist.github.com/jonas054/eb08d4563ed56c88a081). Maybe you could start with that and see if you can make it nicer @thirtysixthspan. The `heredoc_ranges` method comes from `lib/rubocop/cop/mixin/autocorrect_alignment.rb` but is slightly altered and should of course not be duplicated. One example fails with my code but I'm not sure if the spec example is correct.

Comment (bbatsov) [inline lib/rubocop/cop/style/redundant_begin.rb]: @thirtysixthspan ping :-)

**Comment by thirtysixthspan** [comment]
The Parse library returns incorrect source for heredocs which affected the RedundantBegin cop.
```
require 'astrolabe/builder'
require 'parser/current'

buffer = Parser::Source::Buffer.new('(string)')
buffer.source = <<-CODE
puts <<-EOS
  This is a string
EOS
CODE
builder = Astrolabe::Builder.new
parser = Parser::CurrentRuby.new(builder)
root_node = parser.parse(buffer)

puts buffer.source
# puts <<-EOS
#  This is a string
# EOS

puts root_node.loc.expression.source
# puts <<-EOS
```
I've updated this PR with a work around that takes advantage of the fact that the range for an individual node that is a heredoc, will correctly report the last line of the heredoc.

**Comment by bbatsov** [comment]
If there's a bug in `parser` you should submit to @whitequark's attention. He's like the Flash when it comes to fixing bugs. :-)

**Comment by whitequark** [comment]
While I did find an unrelated bug as a result of this report, this is not actually a bug, in that it is [documented](https://whitequark.github.io/parser/file.AST_FORMAT.html#here-document) behavior:
```
puts <<-EOS
~~~~ selector
~~~~~~~~~~~ expression
(str "  This is a string\n")
puts <<-EOS
     ~~~~~~ expression
  This is a string
~~~~~~~~~~~~~~~~~~... heredoc_body
EOS
~~~ heredoc_end
```

The reason it behaves like that is that it would be prohibitively expensive to rewrite AST::Builder to also check all the intermediate nodes and not just first and last when computing locations. (I am not excited about the prospect of testing that.) However, if doing it in the AST consumers is just as expensive, it might be worth doing something about it.

**Comment by thirtysixthspan** [comment]
Here is another example of valid (but poorly styled) Ruby where heredocs are not respected by the auto-correction
```
# encoding: utf-8

p = proc do 
  |a: <<-A_DEFAULT, b: <<-B_DEFAULT|
    default for A
  A_DEFAULT
    default for B
  B_DEFAULT

  [a, b]
end

puts p.call.to_s
#=> ["    default for A\n", "    default for B\n"]
```
```
rubocop --only MultilineBlockLayout -a test.rb

Offenses:

test.rb:3:17: F: unterminated string meets end of file
p = proc do |a: <<-A_DEFAULT, b: <<-B_DEFAULT|[a, b]
                ^
test.rb:4:3: C: [Corrected] Block argument expression is not on the same line as the block start.
  |a: <<-A_DEFAULT, b: <<-B_DEFAULT|
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
and results in invalid Ruby:
```
# encoding: utf-8

p = proc do |a: <<-A_DEFAULT, b: <<-B_DEFAULT|[a, b]
end

puts p.call.to_s
```
because, I believe, there are similar range issues.

**Comment by whitequark** [comment]
Yeah. I'd like to hear from @bbatsov on how to deal with this. IIRC Rubocop uses @yujinakayama's astrolabe or something, which is a custom AST builder? So even if I have provided an AST transform that fixups the `heredoc_end` into `expression`, that would not help much.

**Comment by thirtysixthspan** [comment]
@bbatsov thoughts on how to proceed?

**Comment by bbatsov** [comment]
> Yeah. I'd like to hear from @bbatsov on how to deal with this. IIRC Rubocop uses @yujinakayama's astrolabe or something, which is a custom AST builder?

@whitequark Yes, we use https://github.com/yujinakayama/astrolabe. It's a pretty thin wrapper over your `ast` gem, so I don't really understand what your concern about it.

**Comment by whitequark** [comment]
@bbatsov Ah, I see now, that should not be a problem then.

**Comment by bbatsov** [comment]
@whitequark On a related note - you might have a look at astrolabe and add some of its enhancements to `ast` itself. I believe they'll make the lives of `ast`'s users a bit nicer. This functionality eliminated a ton of boilerplate and custom functionality from RuboCop.

**Comment by whitequark** [comment]
@bbatsov If I recall correctly one of astrolabe's main features is that it makes ASTs mutable, and making them immutable was a deliberate design choice.

**Comment by bbatsov** [comment]
> @bbatsov If I recall correctly one of astrolabe's main features is that it makes ASTs mutable, and making them immutable was a deliberate design choice.

I'm not sure. I'll have to consult the astrolabe code. Btw, @yujinakayama, feel free to join us here.

**Comment by thirtysixthspan** [comment]
@bbatsov Can I suggest a merge at this point? That way we, at a minimum, we can avoid heredoc problems for the time being. It is not looking like we will see upstream changes to address this in the near future. Otherwise, can you provide guidance on how best to proceed?

**Comment by bbatsov** [comment]
@thirtysixthspan Fine with me. Squash the commits, align their messages to our contribution guidelines and update the changelog.

**Comment by bbatsov** [comment]
@thirtysixthspan ping

**Comment by bbatsov** [comment]
@thirtysixthspan ping 2 :-)

**Comment by bbatsov** [comment]
Closing due to lack of activity.

**Pull request closed without merging**